### PR TITLE
[BEAM-3159] Deprecated DoFnTester

### PR DESCRIPTION
**Pull request by kennknowles**
Deprecates DoFnTester and removes from the core SDK tests.

----

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
run java gradle precommit

**Comment by kennknowles** [comment]
Looked like flakes in IOs but only on gradle build.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 1]
```diff
@@ -51,56 +51,34 @@
...
- * List<OutputT> testOutputs = fnTester.processBundle(testInput);
- * Assert.assertThat(testOutputs, Matchers.hasItems(...));
- *
- * // Process a bigger bundle:
- * Assert.assertThat(fnTester.processBundle(i1, i2, ...), Matchers.hasItems(...));
- * } </pre>
- *
- * @param <InputT> the type of the {@link DoFn}'s (main) input elements
- * @param <OutputT> the type of the {@link DoFn}'s (main) output elements
+ * @deprecated Use {@link TestStream} with the {@code DirectRunner}.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Mention `TestPipeline` and `PAssert` too? Maybe link to the Beam testing guide https://beam.apache.org/contribute/testing/#effective-use-of-the-testpipeline-junit-rule .

Also, I think mentioning `TestStream` is only needed if the user is testing something with nontrivial timestamps/watermarks - if not, they can just use Create.of().

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/CoGroupByKeyTest.java [hunk 2]
```diff
@@ -424,19 +428,31 @@ public void testConsumingDoFn() throws Exception {
...
+                ImmutableList.of(purchasesTag, addressesTag, namesTag)),
+            UnionCoder.of(
+                ImmutableList.of(
+                    StringUtf8Coder.of(),
+                    StringUtf8Coder.of(),
+                    StringUtf8Coder.of()))));
+
+    PCollection<KV<String, Integer>> results =
+        p.apply(
+                TestStream.create(coder)
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/CoGroupByKeyTest.java]: Hm why not just `Create.of()` here too?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 1 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/CoGroupByKeyTest.java (hunk 2 above)
Comment (kennknowles) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/CoGroupByKeyTest.java]: Fair point. I was being pretty mechanical in my conversion. Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 3]
```diff
@@ -51,56 +53,37 @@
...
-   * {@link BoundedWindow}.
+   * @deprecated Use {@link TestStream} with the {@code DirectRunner}.
    */
   @SuppressWarnings("unchecked")
+  @Deprecated
   public static <InputT, OutputT> DoFnTester<InputT, OutputT> of(DoFn<InputT, OutputT> fn) {
     checkNotNull(fn, "fn can't be null");
+    LOG.warn(
+        "Your tests use DoFnTester, which may not exercise DoFns correctly. "
+            + "Please use TestStream instead.");
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: I'd suggest to rephrase all "use TestStream" in this PR to "use TestPipeline".

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 4]
```diff
@@ -57,11 +58,8 @@
  * serializability, lack of access to global shared mutable state,
  * requirements for failure tolerance, and benefits of optimization.
  *
- * <p>{@code DoFn}s can be tested in a particular
- * {@code Pipeline} by running that {@code Pipeline} on sample input
- * and then checking its output.  Unit testing of a {@code DoFn},
- * separately from any {@code ParDo} transform or {@code Pipeline},
- * can be done via the {@link DoFnTester} harness.
+ * <p>{@link DoFn DoFns} can be tested by using {@link TestStream}
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: by using TestPipeline

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 5]
```diff
@@ -51,56 +53,37 @@
...
- * // Process a bigger bundle:
- * Assert.assertThat(fnTester.processBundle(i1, i2, ...), Matchers.hasItems(...));
- * } </pre>
- *
- * @param <InputT> the type of the {@link DoFn}'s (main) input elements
- * @param <OutputT> the type of the {@link DoFn}'s (main) output elements
+ * @deprecated Instead of this, write a {@link TestPipeline} that verifies expectations using {@link
+ *     PAssert} and run it with the {@code DirectRunner} (or other runners, as applicable). For
+ *     simple cases, use {@link Create} to provide input. For complex cases involving triggers, use
+ *     {@link TestStream} for input data and behavior.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: I'm actually not sure if TestStream is relevant at all to testing of DoFn's. What TestStream simulates, affects GBKs and Combine's, not ParDo's.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: It does enforce bundle boundaries, though I'm not sure it is required to do so.

Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Hm it's an interesting point, but I'm not sure of that either; or at least I've never seen a test that relies on that, so it's not a common test pattern (testing for particular cases of bundle boundaries is also pretty rare) and I still would suggest to recommend people use TestPipeline. If it turns out that TestStream is useful for testing bundle boundaries, then make a note about that too?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java (hunk 4 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 5 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: I didn't port IOs that use `DoFnTester`. I know they are interested in bundle boundaries. I've moved it to the `DoFn` javadoc and rephrased to be quite explicit. WDYT?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 3 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Agreed and done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 6]
```diff
@@ -48,24 +50,21 @@
...
- * <p>{@code DoFn}s can be tested in a particular
- * {@code Pipeline} by running that {@code Pipeline} on sample input
- * and then checking its output.  Unit testing of a {@code DoFn},
- * separately from any {@code ParDo} transform or {@code Pipeline},
- * can be done via the {@link DoFnTester} harness.
+ * <p>{@link DoFn DoFns} can be tested by using {@link TestPipeline}. You can verify their
+ * functional correctness in a local test using the {@code DirectRunner} as well as running
+ * integration tests with your production runner of choice. To create a bundle boundary between
+ * input elements, for testing {@StartBundle} and {@link FinishBundle}, you can use {@link
+ * TestStream} to force quiescence at selected points of your input.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: I'm not sure what "force quiescence" means. It seems that in direct runner indeed 1 `addElements` call generates exactly 1 bundle, at least for the immediately following DoFn, so maybe say "Typically you can generate the input data using Create.of() or other transforms. However, if you need to test behavior of StartBundle and FinishBundle with particular deterministic bundle boundaries, you can use TestStream, where each call to addElements produces 1 bundle", and document that on TestStream#addElements.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: The semantics of `TestStream` are that all computable consequences of each builder method call must be fully evaluated before going on to the next. So if you `addElements` definitely all of those elements must be fully processed and committed. I admit this is also a potentially undefined notion. But I would say the problem is the spec - certainly the bundle has to be finished and really the sinks should have written everything that is ready based on those elements.

I think each call to `addElements` produces at least one bundle - could be 1 bundle per element. But no element can be left hanging in an unfinished bundle.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: I'll rephrase to something like what you've said before merging.

Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Yeah it's icky: if you pass TestStream through an identity ParDo and direct runner chose to materialize the results of that, these guarantees all would be bogus. I'm honestly not sure if it's even a good idea to document it anymore. Up to you.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Ah, good point. TBH if we can't force bundle boundaries I'm a little sad about this change. I dislike `DoFnTester` a lot but we do need real tests of the IO DoFns. Still, ITs are probably better.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Wait, actually that is not true. Materialization does not indicate quiescence. I believe they are required to be processed unless blocked by watermark/trigger progress.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: I went with your phrasing, cutting off just the last phrase about exactly how it works. Hoping the lack of detail here implies you had better read the docs.

**Comment by kennknowles** [comment]
Rebased, squashed, fixed javadoc link error. Will get to green.

**Comment by kennknowles** [comment]
run java gradle precommit

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
run java gradle precommit

**Pull request merged**